### PR TITLE
contracts/swap, swap: verify result of submitChequeBeneficiary

### PR DESCRIPTION
**Pull request by Eknir**
We make the functionality of submitChequeBeneficiary to return a txHash and two channels on which we can listen to a receipt and error. 
This pattern is in line with [web3 PromiEvents](https://web3js.readthedocs.io/en/1.0/callbacks-promises-events.html).

I am aware that there are no tests yet. If possible, I woud like to have help with this.

### CODE REVIEW

File: swap/peer.go [hunk 1]
```diff
@@ -195,14 +196,25 @@ func (sp *Peer) handleEmitChequeMsg(ctx context.Context, msg interface{}) error
...
-		log.Error(fmt.Sprintf("error while calling submit cheque beneficiary: %s", err.Error()))
+		log.Info("Could not get an instance of simpleSwap")
+		return err
+	}
+	tx, receipt, err := otherSwap.SubmitChequeBeneficiary(opts, sp.backend, big.NewInt(int64(cheque.Serial)), big.NewInt(int64(cheque.Amount)), big.NewInt(int64(cheque.Timeout)), cheque.Sig)
+	log.Info(fmt.Sprintf("Got tx: %x", tx))
+	select {
+	case err := <-err:
+		if err != nil {
+			log.Error("Got error when calling submitChequeBeneficiary: ", err)
```
Comment (mortelli) [inline swap/peer.go:210]: In this case, `err` should be the return value of the function.

File: contracts/swap/swap.go [hunk 2]
```diff
@@ -89,12 +89,23 @@ func (s *Swap) ContractParams() *Params {
...
-// SubmitChequeBeneficiary is used to cash in a cheque
-func (s *Swap) SubmitChequeBeneficiary(opts *bind.TransactOpts, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Transaction, error) {
-	return s.Instance.SubmitChequeBeneficiary(opts, serial, amount, timeout, ownerSig)
+// SubmitChequeBeneficiary is used to cash in a cheque. It returns directly the transactionHash, a channel on which can be listened for the transaction to be mined and a channel where we can send an error
+func (s *Swap) SubmitChequeBeneficiary(auth *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (common.Hash, <-chan *types.Receipt, <-chan error) {
+	errorCh := make(chan error)
+	receiptCh := make(chan *types.Receipt)
+	tx, err := s.Instance.SubmitChequeBeneficiary(auth, serial, amount, timeout, ownerSig)
+	if err != nil {
+		errorCh <- err
```
Comment (ralph-pichler) [inline contracts/swap/swap.go]: if `err != nil` the following go routine should never be executed

File: contracts/swap/swap.go [hunk 3]
```diff
@@ -89,12 +89,23 @@ func (s *Swap) ContractParams() *Params {
...
 }
 
-// SubmitChequeBeneficiary is used to cash in a cheque
-func (s *Swap) SubmitChequeBeneficiary(opts *bind.TransactOpts, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (*types.Transaction, error) {
-	return s.Instance.SubmitChequeBeneficiary(opts, serial, amount, timeout, ownerSig)
+// SubmitChequeBeneficiary is used to cash in a cheque. It returns directly the transactionHash, a channel on which can be listened for the transaction to be mined and a channel where we can send an error
+func (s *Swap) SubmitChequeBeneficiary(auth *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (common.Hash, <-chan *types.Receipt, <-chan error) {
+	errorCh := make(chan error)
+	receiptCh := make(chan *types.Receipt)
+	tx, err := s.Instance.SubmitChequeBeneficiary(auth, serial, amount, timeout, ownerSig)
```
Comment (ralph-pichler) [inline contracts/swap/swap.go:98]: Should probably already happen in a go routine

File: contracts/swap/swap.go [hunk 4]
```diff
@@ -89,12 +89,23 @@ func (s *Swap) ContractParams() *Params {
...
+		errorCh <- err
+		receiptCh <- receipt
+	}()
+	return tx.Hash(), receiptCh, errorCh
 }
 
-func (s *Swap) InstanceAt(address common.Address, backend bind.ContractBackend) *contract.SimpleSwap {
-	ref, _ := contract.NewSimpleSwap(address, backend)
-	return ref
+func (s *Swap) InstanceAt(address common.Address, backend bind.ContractBackend) error {
```
Comment (vojtechsimetka) [inline contracts/swap/swap.go]: isn't this part of a different PR?

File: contracts/swap/swap.go [hunk 5]
```diff
@@ -89,12 +89,23 @@ func (s *Swap) ContractParams() *Params {
...
+	s.Instance, err = contract.NewSimpleSwap(address, backend)
```
Comment (ralph-pichler) [inline contracts/swap/swap.go:117]: Ok for now but we should really rethink this

File: swap/peer.go [hunk 6]
```diff
@@ -195,14 +196,26 @@ func (sp *Peer) handleEmitChequeMsg(ctx context.Context, msg interface{}) error
...
-	// handling error
-	// asynchronous call to blockchain, might not get error back directly. If we get a txhash directly, we still have to check the result of this tx.
-	ref := sp.swap.contractReference.InstanceAt(cheque.Contract, sp.backend)
-	_, err = ref.SubmitChequeBeneficiary(opts, big.NewInt(int64(cheque.Serial)), big.NewInt(int64(cheque.Amount)), big.NewInt(int64(cheque.Timeout)), cheque.Sig)
+	//TODO: make instanceAt to directly return a swap type
+	otherSwap := swap.New()
+	err := otherSwap.InstanceAt(cheque.Contract, sp.backend)
 	if err != nil {
-		log.Error(fmt.Sprintf("error while calling submit cheque beneficiary: %s", err.Error()))
+		log.Info("Could not get an instance of simpleSwap")
```
Comment (ralph-pichler) [inline swap/peer.go]: should be `log.Error`

File: swap/peer.go [hunk 7]
```diff
@@ -195,14 +196,26 @@ func (sp *Peer) handleEmitChequeMsg(ctx context.Context, msg interface{}) error
...
+		return err
+	}
+	tx, receipt, err := otherSwap.SubmitChequeBeneficiary(opts, sp.backend, big.NewInt(int64(cheque.Serial)), big.NewInt(int64(cheque.Amount)), big.NewInt(int64(cheque.Timeout)), cheque.Sig)
```
Comment (ralph-pichler) [inline swap/peer.go]: make it explicit that receipt and err are channels

File: contracts/swap/swap.go [hunk 8]
```diff
@@ -89,12 +89,23 @@ func (s *Swap) ContractParams() *Params {
...
+func (s *Swap) SubmitChequeBeneficiary(auth *bind.TransactOpts, backend Backend, serial *big.Int, amount *big.Int, timeout *big.Int, ownerSig []byte) (common.Hash, <-chan *types.Receipt, <-chan error) {
+	errorCh := make(chan error)
+	receiptCh := make(chan *types.Receipt)
+	tx, err := s.Instance.SubmitChequeBeneficiary(auth, serial, amount, timeout, ownerSig)
+	if err != nil {
+		errorCh <- err
+	}
+	go func() {
+		receipt, err := bind.WaitMined(auth.Context, backend, tx)
+		errorCh <- err
```
Comment (ralph-pichler) [inline contracts/swap/swap.go]: should not be sent to channel if nil

**Review by ralph-pichler — CHANGES_REQUESTED** [review]

**Review by ralph-pichler — APPROVED** [review]
LGTM now

**Pull request merged**